src/TheOpenNetwork/Signer.cpp

```diff
@@ -0,0 +1,66 @@
+// Copyright © 2017-2022 Trust Wallet.
+//
+// This file is part of Trust. The full Trust copyright notice, including
+// terms governing use, modification, and redistribution, is contained in the
+// file LICENSE at the root of the source code distribution tree.
+
+#include "Signer.h"
+
+#include "Base64.h"
+
+#include "TheOpenNetwork/wallet/WalletV4R2.h"
+#include "Everscale/WorkchainType.h"
+
+namespace TW::TheOpenNetwork {
+
+Data Signer::createTransferMessage(const PublicKey& publicKey, const PrivateKey& privateKey, const Proto::Transfer& transfer) {
+    std::unique_ptr<Wallet> wallet;
+    const int8_t workchainId = Everscale::WorkchainType::Basechain;
+
+    switch (transfer.wallet_version()) {
+    case Proto::WalletVersion::WALLET_V4_R2: {
+        wallet = std::make_unique<WalletV4R2>(publicKey, workchainId);
+        break;
+    }
+    default:
+        throw std::invalid_argument("Unsupported wallet version");
+    }
+
+    const auto msg = wallet->createTransferMessage(
+        privateKey,
+        Address(transfer.dest()),
+        transfer.amount(),
+        transfer.sequence_number(),
+        static_cast<uint8_t>(transfer.mode()),
+        transfer.expire_at(),
+        transfer.comment()
+    );
+
+    Data result{};
+    msg->serialize(result);
+    return result;
+}
+
+Proto::SigningOutput Signer::sign(const Proto::SigningInput &input) noexcept {
+    const auto& privateKey = PrivateKey(input.private_key());
+    const auto& publicKey = privateKey.getPublicKey(TWPublicKeyTypeED25519);
+
+    auto protoOutput = Proto::SigningOutput();
+
+    switch (input.action_oneof_case()) {
+    case Proto::SigningInput::ActionOneofCase::kTransfer: {
+        const auto& transfer = input.transfer();
+
+        try {
```

Comment:
This is done so that the method becomes "noexcept"
There are many possible exceptions, for example, one can write a cell payload that is too long and end up with this:
https://github.com/trustwallet/wallet-core/blob/master/src/Everscale/Cell.cpp#L351

The same logic at the top for `decodeUserFriendlyAddress`
We used this as a reference: https://github.com/trustwallet/wallet-core/blob/master/src/Everscale/Address.cpp#L26
What's the correct approach here?

Reply:
One of the trivial possible exceptions is std::bad_alloc. We need to create a new string (see TW::Base64::encode for example) to get output. I suppose it is common for sign methods for any blockchain

Reply:
About exceptions: we will try to make the code as much as we can without exceptions and handle individual types of exceptions separately

Reply:
the problem is - wasm doesn't catch exceptions, if there's an exception, it will just crash. So try not to use them ;)

We want to support TON from wasm (e.g. for using it in chrome extension in the future). That's why it's important to add wasm unit tests as it will allow us to see problems (e.g. with exceptions).

Reply:
So I see these places with exceptions including everscale directory:
  `PrivateKey` and `PublicKey` constructors, `parse_hex function`, `stoi`, `decodeBase64Url`, `decodeBase64`, `transform_width`, `binary_from_base64`
  Everywhere where there is `throw`
  
  I suppose that validateAddress method should not check address with exceptions. However I see there is a lot of code for example for ethereum which has  `PrivateKey` and `PublicKey` constructors, `parse_hex function`, and a lot of `throw` statements. To which degree should we remove exceptions?

Reply:
> So I see these places with exceptions including everscale directory: `PrivateKey` and `PublicKey` constructors, `parse_hex function`, `stoi`, `decodeBase64Url`, `decodeBase64`, `transform_width`, `binary_from_base64` Everywhere where there is `throw`
> 
> I suppose that validateAddress method should not check address with exceptions. However I see there is a lot of code for example for ethereum which has `PrivateKey` and `PublicKey` constructors, `parse_hex function`, and a lot of `throw` statements. To which degree should we remove exceptions?
> 
> Please not we made isValid for address working without exceptions

Discussed it in the chat, decided to make only "validateAddress" w/o exceptions.